nmt_test.go

```diff
@@ -845,3 +870,272 @@ func exampleTreeWithEightLeaves() *NamespacedMerkleTree {
+		// the ProveNamespace method  should produce an error for any input namespace ID.
+		{"unordered leaf hashes: query for the last namespace", treeWithUnorderedLeafHashes, namespace.ID{8, 8}, true, ErrUnorderedSiblings}, // for a tree with an unordered set of leaves,
+		// the ProveNamespace method  should produce an error for any namespace ID.
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := tt.tree.ProveNamespace(tt.nID)
+			assert.Equal(t, tt.wantErr, err != nil)
+			if tt.wantErr {
+				assert.True(t, errors.Is(err, tt.errType))
+			}
+		})
+	}
+}
+
+// Test_Root_Error tests that the Root method returns an error when the underlying tree is in an invalid state, such as when the leaves are not ordered by namespace ID or when a leaf is corrupt.
+func Test_Root_Error(t *testing.T) {
+	// create an NMT with 8 sequentially namespaced leaves, numbered from 1 to 8.
+	treeWithCorruptLeaf := exampleTreeWithEightLeaves()
+	// corrupt a leaf
+	treeWithCorruptLeaf.leaves[4] = treeWithCorruptLeaf.leaves[4][:treeWithCorruptLeaf.NamespaceSize()-1]
+
+	// create an NMT with 8 sequentially namespaced leaves, numbered from 1 to 8.
+	treeWithUnorderedLeaves := exampleTreeWithEightLeaves()
+	// swap the positions of the 4th and 5th leaves
+	swap(treeWithUnorderedLeaves.leaves, 4, 5)
+
+	tests := []struct {
+		name    string
+		tree    *NamespacedMerkleTree
+		wantErr bool
+		errType error
+	}{
+		{"corrupt leaf hash", treeWithCorruptLeaf, true, ErrInvalidLeafLen},
+		{"unordered leaf hashes", treeWithUnorderedLeaves, true, ErrUnorderedSiblings},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := tt.tree.Root()
+			assert.Equal(t, tt.wantErr, err != nil)
+			if tt.wantErr {
+				assert.True(t, errors.Is(err, tt.errType))
+			}
+		})
+	}
+}
+
+// Test_computeRoot_Error tests that the Root method returns an error when the underlying tree is in an invalid state, such as when the leaves are not ordered by namespace ID or when a leaf is corrupt.
+func Test_computeRoot_Error(t *testing.T) {
+	// create an NMT with 8 sequentially namespaced leaves, numbered from 1 to 8.
+	treeWithCorruptLeaf := exampleTreeWithEightLeaves()
+	// corrupt a leaf
+	treeWithCorruptLeaf.leaves[4] = treeWithCorruptLeaf.leaves[4][:treeWithCorruptLeaf.NamespaceSize()-1]
+
+	// create an NMT with 8 sequentially namespaced leaves, numbered from 1 to 8.
+	treeWithUnorderedLeaves := exampleTreeWithEightLeaves()
+	// swap the positions of the 4th and 5th leaves
+	swap(treeWithUnorderedLeaves.leaves, 4, 5)
+
+	tests := []struct {
+		name    string
+		tree    *NamespacedMerkleTree
+		wantErr bool
+		errType error
+	}{
+		{"corrupt leaf", treeWithCorruptLeaf, true, ErrInvalidLeafLen},
+		{"unordered leaves", treeWithUnorderedLeaves, true, ErrUnorderedSiblings},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := tt.tree.Root()
```

Comment:
[question] should this invoke computeRoot? It currently seems identical to the above test:

```go
func Test_Root_Error(t *testing.T) {
    // ...

			_, err := tt.tree.Root()
```

Reply:
Thanks for finding this, please see the corrected version in this commit https://github.com/celestiaorg/nmt/pull/136/commits/3fd5dab14ea4b8cb0f3c3046cc07b7af8079a141